stan/math/rev/fun/log_sum_exp.hpp

```diff
@@ -82,8 +51,19 @@ class log_sum_exp_matrix_vari : public op_matrix_vari {
  */
 template <typename T, require_container_st<is_var, T>* = nullptr>
 inline auto log_sum_exp(const T& x) {
-  return apply_vector_unary<T>::reduce(x, [&](const auto& v) {
-    return var(new internal::log_sum_exp_matrix_vari(v.eval()));
+  return apply_vector_unary<T>::reduce(x, [](const auto& v) {
+    const auto& v_ref = to_ref(v);
+    const auto& v_val = value_of(v_ref);
+
+    var res = log_sum_exp(v_val);
+    auto arena_v = to_arena(v_ref);
+
+    reverse_pass_callback([arena_v, res]() mutable {
+      arena_v.adj()
+          += res.adj() * (arena_v.array().val() - res.val()).exp().matrix();
```

Comment:
[optional] You are using value of `arena_v` in both forward and reverse pass. You could store it in arena variable to avoid extracting it from `arena_v` twice. This would probably be faster at expense of some extra memory usage. Same change is also possible in scalar `log_sum_exp`.

Reply:
The benchmarks show it's about 4% slower at worst (https://bbbales2.shinyapps.io/math-shiny-perf/). I'm gonna merge this and then we can do the performance as part of the other pull.